src/numerics/include/metaphysicl/dualnumber_decl.h

```diff
@@ -85,36 +89,50 @@ class DualNumber : public safe_bool<DualNumber<T,D> >
   DualNumber<T,D> operator! () const;
 
   template <typename T2, typename D2>
-  DualNumber<T,D>& operator+= (const DualNumber<T2,D2>& a);
+  DualNumber<T, D> & operator+= (const DualNumber<T2,D2>& a);
 
   template <typename T2>
-  DualNumber<T,D>& operator+= (const T2& a);
+  DualNumber<T, D> & operator+= (const T2& a);
 
   template <typename T2, typename D2>
-  DualNumber<T,D>& operator-= (const DualNumber<T2,D2>& a);
+  DualNumber<T, D> & operator-= (const DualNumber<T2,D2>& a);
 
   template <typename T2>
-  DualNumber<T,D>& operator-= (const T2& a);
+  DualNumber<T, D> & operator-= (const T2& a);
 
   template <typename T2, typename D2>
-  DualNumber<T,D>& operator*= (const DualNumber<T2,D2>& a);
+  DualNumber<T, D> & operator*= (const DualNumber<T2,D2>& a);
 
   template <typename T2>
-  DualNumber<T,D>& operator*= (const T2& a);
+  DualNumber<T, D> & operator*= (const T2& a);
 
   template <typename T2, typename D2>
-  DualNumber<T,D>& operator/= (const DualNumber<T2,D2>& a);
+  DualNumber<T, D> & operator/= (const DualNumber<T2,D2>& a);
 
   template <typename T2>
-  DualNumber<T,D>& operator/= (const T2& a);
+  DualNumber<T, D> & operator/= (const T2& a);
+
+  operator T &()
+  {
+    DoOnce(std::cout << "This message will only appear once: You are operating on a "
+                        "DualNumber value without operating on its derivatives. "
+                        "This may have a deleterious impact on your Jacobian calculations!\n");
```

Comment:
I definitely sympathize with the conflict between "sometimes we need to pull out a raw value for compatibility with a non-generic algorithm" and "pulling out a raw value unintentionally is horribly unsafe", but this can't be the right solution.

If application code uses this method safely, a false positive warning on cout will make attentive users ignore your warnings.

If application code uses this method unsafely, unattentive users (i.e. most users all of the time and all users most of the time) may miss the warning.

If application code uses this method unsafely on some-but-not-all code paths (say, in an MPI-based code that disables cout on ranks above 0, *like MOOSE*) then even attentive users might see clean output from one run only to unexpectedly get incorrect derivatives from another run.

At the very least this code needs to be wrapped with some sort of ```#ifdef METAPHYSICL_ENABLE_HORRIBLE_CONVERSIONS``` non-default configure option.

More likely, there's a much better solution possible, but I may simply not understand the underlying problem.  I'd have guessed that you want to call ```legacy_code(input.value())``` when input is a DualNumber, but that doesn't compile when input is raw floating point?  That's what ```legacy_code(raw_value(input))``` would fix, though you'd have to add an identity ```raw_value``` definition in the ```#ifndef HAVE_METAPHYSICL``` case for a MetaPhysicL-optional application.

Whatever the application is, the person who can tell you whether your code is making a deleterious mistake is the application programmer, not the application user.  In that case the right thing to make the programmer do has to be *something* like "use raw_value to prove you're being serious", not "hope you never invoke a destructive implicit downconversion unintentionally..."

... I tried not to be an ass or even to just sound like an ass in this comment, but I think I'm 0 for 2, for which I apologize.

Reply:
You're right that this is not the right solution.

The most common situation where we are going to run into these conversions is:
```c++
SomeMaterial::computeProperties
{
  :
  :
  :
  _some_mat_prop = a * b - c + d
}
```
where some/all of `a,...,d` are `DualNumbers` declared in some base class by a "DualNumber-aware" application developer, but `_some_mat_prop` is a raw number declared by a downstream "DualNumber-oblivious" developer. I guess I need to think on this more as to how best to tackle this scenario. We had a debate in the cube a couple weeks ago about this and didn't come to any great solution.

Reply:
```_some_mat_prop = raw_value(a * b - c + d)```

(or possibly ```MetaPhysicL::raw_value``` if argument-dependent-lookup isn't playing nicely)

The key is whether ```SomeMaterial``` is written by a "DualNumber-aware" developer, I think.  If the answer will always be "yes" then you should be in good shape, but if you're pulling in an unmodified ```generic_some_material_code.h``` from a developer who wrote templated generic algorithms but whose psychic abilities did not anticipate a "raw_value" API then you're out of luck.